adapters/ads_adapter.py

```diff
@@ -1,22 +1,167 @@
+Subscribes to Beckhof ADS variables and sends it to the GUI tool.
+
+Auto subscribe only works for DINT and LREAL for now.
+
+Example usage:
+    python ads_adapter.py --regex GVl.*temperature
 
 """
+from lognplot.client import LognplotTcpClient
+import pyads
+import time
+import ctypes
+import struct
+import datetime
+import re
+import argparse
+from collections import namedtuple
+from typing import Type
 
-# TODO: implement.
 
+class AdsClient:
+    Entry = namedtuple(
+        "Entry", ["name", "typename", "comment", "datatype", "datatype_size"]
+    )
 
+    # Some data types have same number, maybe use type string field found in __parse_ads_entries...
+    DATATYPE_MAP = {
+        17: pyads.PLCTYPE_BOOL,  # ADST_UINT8
+        17: pyads.PLCTYPE_BYTE,  # ADST_UINT8
+        3: pyads.PLCTYPE_DINT,  # ADST_INT32
+        19: pyads.PLCTYPE_DWORD,  # ADST_UINT32
+        2: pyads.PLCTYPE_INT,  # ADST_UINT16
+        5: pyads.PLCTYPE_LREAL,  # ADST_REAL64
+        4: pyads.PLCTYPE_REAL,  # ADST_REAL32
+        16: pyads.PLCTYPE_SINT,  # ADST_INT8
+        19: pyads.PLCTYPE_UDINT,  # ADST_UINT32
+        18: pyads.PLCTYPE_UINT,  # ADST_UINT16
+        17: pyads.PLCTYPE_USINT,  # ADST_UINT8
+        18: pyads.PLCTYPE_WORD,  # ADST_UINT16
+    }  # type: Dict[str, Type]
 
-from lognplot.client import LognplotTcpClient
+    def __init__(self, ams_net_id: str, ams_net_port: str):
+        self.notification_handles = []
+        self.plc = pyads.Connection(ams_net_id, ams_net_port)
+        self.plc.open()
+
+    def __del__(self):
+        for handle in self.notification_handles:
+            self.plc.del_device_notification(*handle)
+
+        self.plc.close()
+
+    def start_logging(self, name: str, plc_type):
+        attr = pyads.NotificationAttrib(ctypes.sizeof(plc_type))
+        handles = self.plc.add_device_notification(
+            name, attr, self.callback(plc_type), pyads.ADSTRANS_SERVERCYCLE
+        )
+        self.notification_handles.append(handles)
+
+    def subscribe(self, pattern: str):
+        parsed_entries = self.get_ads_entries()
+
+        for entry in parsed_entries:
+            if re.match(pattern, entry.name) != None:
+                self.log(entry.name, self.DATATYPE_MAP[entry.datatype])
+
+    def callback(self, plc_type):
+        @self.plc.notification(plc_type)
+        def decorated_callback(handle, name, timestamp, value):
+            client.send_sample(
+                name, self.__format_timestamp(timestamp), float(value),
+            )
+
+        return decorated_callback
+
+    def get_ads_entries(self):
+        upload_info = self.__get_upload_info()
+        entries = self.plc.read(
+            pyads.constants.ADSIGRP_SYM_UPLOAD, 0, ctypes.c_ubyte * upload_info.nSymSize
+        )
+
+        parsed_entries = self.__parse_ads_entries(
+            bytes(entries), upload_info.nSymSize, upload_info.nSymbols
+        )
+        return parsed_entries
+
+    def __format_timestamp(timestamp: datetime):
```

Comment:
This method could be a function or a staticmethod, since self is not used.

Btw, using a double-underscore has some implications with inheritance, I suggest to use a single-underscore to mark the function as "private".